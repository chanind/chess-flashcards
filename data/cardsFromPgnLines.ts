import Flashcard from '../models/Flashcard';
import { Orientation } from './types';
import { Chess } from 'chess.js';

interface CardsFromPgnLinesOptions {
  orientation?: Orientation;
}

class PGNParseError extends Error {}

const hashCard = (card: Flashcard): string => JSON.stringify(card);

/**
 * Helper to turn a list of PGN strings into cards for openings
 * @param pgnLines array of PGN strings
 * @param options object containing optional `orientation`, default 'white'
 * @returns array of Flashcard objects
 */
const cardsFromPgnLines = (
  pgnLines: string[],
  options: CardsFromPgnLinesOptions = {},
): Flashcard[] => {
  const seenCards: Set<string> = new Set();
  const orientation: Orientation = options.orientation ?? 'white';
  const cards: Flashcard[] = [];
  for (const line of pgnLines) {
    const chess = new Chess();
    if (!chess.load_pgn(line)) {
      throw new PGNParseError(`Unable to parse PGN lines: ${line}`);
    }
    const gameReplay = new Chess();
    const previousMoves: string[] = [];
    for (const move of chess.history()) {
      const isPlayerMove = gameReplay.turn() === orientation[0];
      if (isPlayerMove) {
        const card: Flashcard = {
          previousMoves: [...previousMoves],
          correctMove: move,
        };
        const cardHash = hashCard(card);
        // make sure the card hasn't already been added from another line
        if (!seenCards.has(cardHash)) cards.push(card);
        seenCards.add(cardHash);
      }
      previousMoves.push(move);
      gameReplay.move(move);
    }
  }

  return cards;
};

export default cardsFromPgnLines;
