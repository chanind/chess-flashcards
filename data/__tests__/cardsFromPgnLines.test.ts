import cardsFromPgnLines from '../cardsFromPgnLines';

describe('cardsFromPgnLines', () => {
  it('turns PGN lines into flashcards', () => {
    const lines = [
      `1. e4 e5 2. Nc3 Nf6 3. f4 exf4 4. e5 Ng8 5. Nf3 d6 6. d4 dxe5 7. Qe2 Bb4 8. Qxe5+ Qe7 9. Bxf4 *`,
    ];
    const cards = cardsFromPgnLines(lines, { orientation: 'white' });
    expect(cards).toHaveLength(9);
    expect(cards).toMatchSnapshot();
  });

  it('dedupes identical cards from multiple lines', () => {
    const lines = [`1. e4 e5 2. Nc3 Nf6`, `1. e4 e5 2. Nf3 Nf6`];
    const cards = cardsFromPgnLines(lines, { orientation: 'white' });
    expect(cards).toHaveLength(3);
  });
});
