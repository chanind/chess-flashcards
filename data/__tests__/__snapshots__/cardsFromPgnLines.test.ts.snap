// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cardsFromPgnLines turns PGN lines into flashcards 1`] = `
Array [
  Object {
    "correctMove": "e4",
    "previousMoves": Array [],
  },
  Object {
    "correctMove": "Nc3",
    "previousMoves": Array [
      "e4",
      "e5",
    ],
  },
  Object {
    "correctMove": "f4",
    "previousMoves": Array [
      "e4",
      "e5",
      "Nc3",
      "Nf6",
    ],
  },
  Object {
    "correctMove": "e5",
    "previousMoves": Array [
      "e4",
      "e5",
      "Nc3",
      "Nf6",
      "f4",
      "exf4",
    ],
  },
  Object {
    "correctMove": "Nf3",
    "previousMoves": Array [
      "e4",
      "e5",
      "Nc3",
      "Nf6",
      "f4",
      "exf4",
      "e5",
      "Ng8",
    ],
  },
  Object {
    "correctMove": "d4",
    "previousMoves": Array [
      "e4",
      "e5",
      "Nc3",
      "Nf6",
      "f4",
      "exf4",
      "e5",
      "Ng8",
      "Nf3",
      "d6",
    ],
  },
  Object {
    "correctMove": "Qe2",
    "previousMoves": Array [
      "e4",
      "e5",
      "Nc3",
      "Nf6",
      "f4",
      "exf4",
      "e5",
      "Ng8",
      "Nf3",
      "d6",
      "d4",
      "dxe5",
    ],
  },
  Object {
    "correctMove": "Qxe5+",
    "previousMoves": Array [
      "e4",
      "e5",
      "Nc3",
      "Nf6",
      "f4",
      "exf4",
      "e5",
      "Ng8",
      "Nf3",
      "d6",
      "d4",
      "dxe5",
      "Qe2",
      "Bb4",
    ],
  },
  Object {
    "correctMove": "Bxf4",
    "previousMoves": Array [
      "e4",
      "e5",
      "Nc3",
      "Nf6",
      "f4",
      "exf4",
      "e5",
      "Ng8",
      "Nf3",
      "d6",
      "d4",
      "dxe5",
      "Qe2",
      "Bb4",
      "Qxe5+",
      "Qe7",
    ],
  },
]
`;
